### PR TITLE
build docs and push to gh-pages on nightly

### PR DESCRIPTION
**Pull request by mattip**
Copy pytorch/audio#1100 and friends to here, which 
- builds docs on every PR
- pushes the built docs on nightly (to `gh-pages/master`) and when the repo is tagged (to `gh-pages/tag`)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -467,6 +467,63 @@ jobs:
...
+    steps:
+      - attach_workspace:
+          at: ~/workspace
+      - checkout
+      - run:
+          name: install binaries
+          command: |
+            set -x
+            source /usr/local/etc/profile.d/conda.sh && conda activate python${PYTHON_VERSION}
+            pip install $(ls ~/workspace/torchtext*.whl) --pre -f "https://download.pytorch.org/whl/${UPLOAD_CHANNEL}/cpu/torch_${UPLOAD_CHANNEL}.html"
```
Comment (mattip) [inline .circleci/config.yml:484]: This is done via a script file in pytorch/audio, here I copied the smoke_test_linux_pip style.

**Comment by mattip** [comment]
Hmm. Any idea why the same pattern used in the `smoke_test_linux_pip` step is failing to find `/usr/local/etc/profile.d/conda.sh` ? Here is the log https://app.circleci.com/pipelines/github/pytorch/text/1987/workflows/e7d60894-825b-4687-a0e3-782237068949/jobs/55761

**Comment by mthrok** [comment]
> Hmm. Any idea why the same pattern used in the `smoke_test_linux_pip` step is failing to find `/usr/local/etc/profile.d/conda.sh` ? Here is the log https://app.circleci.com/pipelines/github/pytorch/text/1987/workflows/e7d60894-825b-4687-a0e3-782237068949/jobs/55761

The underlying Docker images are different. torchaudio uses `308535385114.dkr.ecr.us-east-1.amazonaws.com/torchaudio/smoke_test:56c846a5-acaa-41a7-92f5-46ec66186c61`, which is build like [this](https://github.com/pytorch/audio/blob/master/.circleci/smoke_test/docker/Dockerfile) and no longer recommended to use.

I think you need to adjust the conda path but I do not think `pytorch/many-linux:100` comes with preinstalled conda. I believe this is the [Dockerfile](https://github.com/pytorch/builder/blob/5862029189311d1808c88789c9c8bd9d276b5e5e/manywheel/Dockerfile).

**Comment by mthrok** [comment]
Well actually `torchtext` has the corresponding [Dockerfile](https://github.com/pytorch/text/blob/master/.circleci/smoke_test/docker/Dockerfile). But apparently torchtext's CI configuration has no mention to the resulting image.

**Comment by mattip** [comment]
The new build_docs check is working. Once this is merged, the upload_docs check will trigger on the next nightly build.

**Comment by zhangguanheng66** [comment]
Seem the failed test is not relevant. Let me re-run it.

**Review by mthrok — APPROVED** [review]
Looks good. Let's see how it goes.

**Comment by mthrok** [comment]
Re-running nightly

**Comment by zhangguanheng66** [comment]
got an error to upload the doc.
https://app.circleci.com/pipelines/github/pytorch/text/2009/workflows/aba8d110-b2de-4a3b-8c76-f42dcc9311aa/jobs/56629

**Comment by mthrok** [comment]
> got an error to upload the doc.
> https://app.circleci.com/pipelines/github/pytorch/text/2009/workflows/aba8d110-b2de-4a3b-8c76-f42dcc9311aa/jobs/56629

@mattip 

Looks like it was caused by `git pull`, which I do not see in the corresponding script in audio.

https://github.com/pytorch/audio/blob/master/.circleci/build_docs/commit_docs.sh

**Pull request merged**